pkg/engine/executor/parse_logfmt.go

```diff
@@ -0,0 +1,189 @@
+		columnOrder, errorList, hasAnyError = parseRequestedKeys(input, requestedKeys, columnBuilders, allocator)
+	} else {
+		// Dynamic path: discover columns as we parse
+		columnOrder, errorList, hasAnyError = parseAllKeys(input, columnBuilders, allocator)
+	}
+
+	// Build final arrays
+	columns := make([]arrow.Array, 0, len(columnOrder)+2)
+	headers := make([]string, 0, len(columnOrder)+2)
+
+	for _, key := range columnOrder {
+		builder := columnBuilders[key]
+		columns = append(columns, builder.NewArray())
+		headers = append(headers, key)
+		builder.Release()
+	}
+
+	// Add error columns if needed
+	if hasAnyError {
+		errorBuilder := array.NewStringBuilder(allocator)
+		errorDetailsBuilder := array.NewStringBuilder(allocator)
+
+		for _, err := range errorList {
+			if err != nil {
+				errorBuilder.Append("LogfmtParserErr")
+				errorDetailsBuilder.Append(err.Error())
+			} else {
+				errorBuilder.AppendNull()
+				errorDetailsBuilder.AppendNull()
+			}
+		}
+
+		columns = append(columns, errorBuilder.NewArray())
+		columns = append(columns, errorDetailsBuilder.NewArray())
+		headers = append(headers, "__error__", "__error_details__")
+
+		errorBuilder.Release()
+		errorDetailsBuilder.Release()
+	}
+
+	return headers, columns
+}
+
+// parseRequestedKeys handles the fast path when specific keys are requested
+// Pre-creates all column builders and processes lines with known columns
+func parseRequestedKeys(input *array.String, requestedKeys []string, columnBuilders map[string]*array.StringBuilder, allocator memory.Allocator) ([]string, []error, bool) {
+	columnOrder := make([]string, 0, len(requestedKeys))
+	errorList := make([]error, 0, input.Len())
+	hasAnyError := false
+
+	// Pre-create all builders
+	for _, key := range requestedKeys {
+		columnBuilders[key] = array.NewStringBuilder(allocator)
+		columnOrder = append(columnOrder, key)
+	}
+
+	// Process each line
+	for i := 0; i < input.Len(); i++ {
+		line := input.Value(i)
+		parsed, err := TokenizeLogfmt(line, requestedKeys) // Already filtered
+		errorList = append(errorList, err)
+		if err != nil {
+			hasAnyError = true
+		}
+
+		// Append values for all requested keys
+		for _, key := range requestedKeys {
+			if value, ok := parsed[key]; ok {
+				columnBuilders[key].Append(value)
+			} else {
+				columnBuilders[key].AppendNull()
+			}
+		}
+	}
+
+	return columnOrder, errorList, hasAnyError
+}
+
+// parseAllKeys handles the dynamic path when no specific keys are requested
+// Discovers columns dynamically as lines are parsed
+func parseAllKeys(input *array.String, columnBuilders map[string]*array.StringBuilder, allocator memory.Allocator) ([]string, []error, bool) {
+	columnOrder := []string{}
+	errorList := make([]error, 0, input.Len())
+	hasAnyError := false
+
+	for i := 0; i < input.Len(); i++ {
+		line := input.Value(i)
+		parsed, err := TokenizeLogfmt(line, nil) // Get all keys
+		errorList = append(errorList, err)
+		if err != nil {
+			hasAnyError = true
+		}
```

Comment:
nit: Instead of returning an error slice, maybe we want to bake in writing to the error column builder, since it's just another column? 

You'll know whether there were any errors by looking at `builder.Len() - builder.NullN()` to get the number of non-null rows.